### PR TITLE
chore(config): improve config schema output with more precise `unevaluatedProperties` + schema ref flattening

### PR DESCRIPTION
**Pull request by tobz**
This PR is a beefy one, so bear with me...

## Context

As part of the work to support generating a "closed" schema (a schema that disallows unknown fields), we had previously implemented support for emitting `unevaluatedProperties: false` within the schema, via a schema visitor, which took the raw, generated schema and traversed it, figuring out which (sub)schemas should be marked, and which should not.

This logic was fairly simplistic, which led to cases where the logic did not correctly account for the schema structure and either mistakenly applied or mistakenly removed the annotation, leading to some parts of the schema incorrectly disallowing unknown fields.

## Solution

This PR does two primary things:

- adds a new visitor, `InlineSingleUseReferencesVisitor`, for flattening out schema references where the schema being referred to is only referred to once
- refactors the existing visitor for unevaluated properties, `DisallowedUnevaluatedPropertiesVisitor`, to denormalize schema references (by flattening) depending on if the schema reference would cause the referred-to schema to be incorrectly unmarked, thereby allowing unevaluated properties to become allowed

### Inlining single-use schema references

The visitor documentation covers this, but essentially, we generally create one schema definition per Rust type, which can mean a lot of types that are specific to, say, a component, are extracted as a schema definition and then referred to via a schema reference... but only once, from that related component.

This is inefficient because you have to "chase" schema references when looking at the overall schema, but also from an anti-don't-repeat-yourself standpoint: there's no need to make it a definition because it's never reused.

As well, it turns out that it makes things slightly faster in the visitor for disallowing unevaluated properties by not having to consider single-use schema references.

### Refactoring the unevaluated properties visitor to denormalize schema references to avoid clobbering marked schemas

For shorthand, we refer to a "marked" schema as one that's been "marked" with `unevaluatedProperties: false`. As a refresher, the visitor would visit all schema objects in a depth-first search, marking object schemas with `unevaluatedProperties: false`, and then work its way back up the graph, unmarking object schemas as necessary in scenario where the referring schema would have the `unevaluatedProperties: false` marking.

Where this logic fell down is that, in some cases, a single schema definition could be referred to in multiple locations, where one location might dictate that the referred-to schema needs to be unmarked, even if the other references would have left the referred-to schema as-is.

We've refactored the visitor so that it does a pass of the schema before visiting, and builds a "flattening eligibility map", which is a map of referred-to schemas and the schemas which refer to them, and whether or not those referrers would cause the schema to be unmarked. Using this map, we can calculate which schema references should instead be denormalized, or flattened, where we copy the referred-to schema and inline it. This allows us to unmark a given referred-to schema at the point of use, while leaving the other references to the schema untouched.

The code takes some efforts to try and figure out which approach is more efficient: flattening all of the references that would cause the schema to be unmarked, or all of the references that would cause the sceham to not be unmarked.

## Reviewer notes

I've added unit tests to both of these visitors, and I would suggest examining those first to familiarize yourself with what the behavior of each visitor is, and what it's attempting to do.

The overarching principle to remember is that, in JSON Schema, it only makes sense to apply `unevaluatedProperties` to object schemas -- schemas which deal with struct fields, essentially -- but if those object schemas are flattened together, using what we would call "subschema validation", each subschema generally meant to only validate a portion of the input, but during validation, it sees the _entire_ input. This requires us to move the `unevaluatedProperties` annotation from the object schemas themselves to the parent, so that each subschema doesn't inadvertently believe it's gotten "unevaluated properties" just because it's being used with subschema validation.

Beyond that, there's a lot of helper code that was written to try and compartmentalize the various preprocessing steps involved before the visitors perform their main visiting logic. There's stuff like `SchemaReference` and `SchemaScopeStack` that aren't _super_ intuitive, because they bounce between one concept -- a schema reference -- and the idea of a schema reference being "active". Don't hesitate to point these things out if you feel like they're too opaque, but do know that I plan to almost immediately follow this PR up with some refactoring to make those areas more concise and correct.

Additionally, some of the machine-generated base docs have changed. I also plan on investigating that in the aforementioned follow-up PR, but left the change as-is because it's fairly benign.

### CODE REVIEW

File: lib/vector-config/src/schema/visitors/unevaluated.rs [hunk 1]
```diff
@@ -1,10 +1,26 @@
-//use indexmap::IndexMap;
+#![allow(warnings)]
```
Comment (spencergilbert) [inline lib/vector-config/src/schema/visitors/unevaluated.rs:1]: What was warning in this file?

**Review by spencergilbert — APPROVED** [review]
This was a remarkable read 😆 didn't see any problems when I read through it.

I think you'll want to run:

```bash
curl -s -S -L 'https://raw.githubusercontent.com/check-spelling/check-spelling/v0.0.21/apply.pl' |
perl - 'https://github.com/vectordotdev/vector/actions/runs/4640241193/attempts/1'
```
locally on your branch to update the spell checker for your new words.

File: lib/vector-config/src/schema/visitors/unevaluated.rs (hunk 1 above)
Comment (tobz) [reply lib/vector-config/src/schema/visitors/unevaluated.rs:1]: Oh, woops. Looks like I left a bunch of stuff around that I added to reduce friction during development. I'll clean this up.

Comment (tobz) [reply lib/vector-config/src/schema/visitors/unevaluated.rs]: Resolved in 310846b7b, including a bunch of cleanup of dead code, etc, that it was hiding.

**Comment by spencergilbert** [comment]
Current failures should be resolved by merging `master` back in, a regression had been merged earlier.

**Pull request merged**